clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java

```diff
@@ -825,27 +826,23 @@ public void seek(TopicPartition partition, OffsetAndMetadata offsetAndMetadata)
 
     @Override
     public void seekToBeginning(Collection<TopicPartition> partitions) {
-        if (partitions == null)
-            throw new IllegalArgumentException("Partitions collection cannot be null");
-
-        acquireAndEnsureOpen();
-        try {
-            Collection<TopicPartition> parts = partitions.isEmpty() ? subscriptions.assignedPartitions() : partitions;
-            subscriptions.requestOffsetReset(parts, OffsetResetStrategy.EARLIEST);
-        } finally {
-            release();
-        }
+        seek(partitions, OffsetResetStrategy.EARLIEST);
     }
 
     @Override
     public void seekToEnd(Collection<TopicPartition> partitions) {
+        seek(partitions, OffsetResetStrategy.LATEST);
+    }
+
+    private void seek(Collection<TopicPartition> partitions, OffsetResetStrategy offsetResetStrategy) {
         if (partitions == null)
             throw new IllegalArgumentException("Partitions collection cannot be null");
 
         acquireAndEnsureOpen();
         try {
-            Collection<TopicPartition> parts = partitions.isEmpty() ? subscriptions.assignedPartitions() : partitions;
-            subscriptions.requestOffsetReset(parts, OffsetResetStrategy.LATEST);
+            Timer timer = time.timer(defaultApiTimeoutMs);
+            ResetOffsetEvent event = new ResetOffsetEvent(partitions, offsetResetStrategy, calculateDeadlineMs(timer));
+            applicationEventHandler.addAndGet(event);
```

Comment:
one concern I had here is that both seek used to directly throw `IllegalStateException` in the app thread if the partitions were not in the assignment, but with this PR that error will now come from the background via an event so could be wrapped I thought. I checked and seems we're safe because of how we do `ConsumerUtils.getResults`:
https://github.com/apache/kafka/blob/3f2cb55a759c562f0ccbfc89fb54f395caf45d9e/clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerUtils.java#L240-L241

Still, it would probably be good to have a simple integration test that would testSeekThrowsIllegalStateIfPartitionsNotAssigned:
`assertThrows(classOf[IllegalStateException], () => consumer.seek(Collections.singletonList(tp))) ` (with tp not assigned). What do you think?

(We have a similar test for [assign](https://github.com/apache/kafka/blob/3f2cb55a759c562f0ccbfc89fb54f395caf45d9e/core/src/test/scala/integration/kafka/api/PlaintextConsumerCallbackTest.scala#L37) that helped catch a regression recently)